### PR TITLE
Add embed compiler

### PR DESCRIPTION
**Pull request by sebastianbenz**
Provides an API to generate embeddable samples including source code and
live preview.

### CODE REVIEW

**Comment by sebastianbenz** [comment]
@pbakaus PTAL - in particular `templates/embed/embed.js`
@kul3r4 FYI

**Comment by kul3r4** [comment]
Just as info, how this will be used? Could you provide an example of code calling this API?

File: lib/Preview.js [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  return DocumentParser.parse(inputString);
+}
+
+/**
+ * Opens the embed html file in a browser to determine the initial height.
+ */
+function generateTemplate(context) {
+  let _page;
+  let _phantom;
+  phantom.create([], { logLevel: 'error' }).then(function (ph) {
```
Comment (pbakaus) [inline lib/Preview.js]: Why not headless Chrome?

File: templates/embed/embed.html [hunk 2]
```diff
@@ -0,0 +1,26 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width,minimum-scale=1">
+    <link href="https://fonts.googleapis.com/css?family=Roboto:100,300,400" rel="stylesheet">
```
Comment (pbakaus) [inline templates/embed/embed.html]: Can we change this to system fonts?

File: templates/embed/embed.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+ * Post a message containing the document height to the amp-iframe to trigger the resize.
+ */
+function postEmbedHeightToViewer() {
+  var newHeight = document.getElementById('tabinterface').clientHeight;
+  if (newHeight < MIN_IFRAME_HEIGHT) {
+    console.log('embed height too small, reset height to 100px');
+    newHeight = MIN_IFRAME_HEIGHT;
+  }
+  // Tell the viewer about the new size
+  window.parent.postMessage({
```
Comment (pbakaus) [inline templates/embed/embed.js]: Didn't even know about this ability, super cool!

Comment (pbakaus) [inline templates/embed/embed.js]: Is this documented anywhere? If not, should I?

File: templates/embed/embed.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    setTimeout(fitPreviewIntoAvailableSpace, 200);
+  }
+}
+
+/**
+ * Resize the preview based on the available width and the resulting preview content height.
+ */
+function fitPreviewIntoAvailableSpace() {
+  // 1. Change width of preview window to let the preview resize itself
+  setPreviewDimension(document.documentElement.clientWidth, preview.height);
```
Comment (pbakaus) [inline templates/embed/embed.js]: Why do we need JS for that? Can't this step be done in CSS?

File: templates/embed/embed.js (hunk 3 above)
Comment (sebastianbenz) [inline templates/embed/embed.js]: It is:

https://www.ampproject.org/docs/reference/components/amp-iframe#iframe-resizing
https://ampbyexample.com/components/amp-iframe/#resizable-iframes

Comment (pbakaus) [inline templates/embed/embed.js]: Ah excellent, thanks!

File: templates/embed/embed.html (hunk 2 above)
Comment (sebastianbenz) [inline templates/embed/embed.html]: I've added it because ampproject.org uses Roboto.

Comment (pbakaus) [inline templates/embed/embed.html]: Fair point. I'm likely going to remove soon. Can revisit as soon as I've removed it.

**Comment by sebastianbenz** [comment]
@kul3r4 there is `lib/example/generate.js`

File: templates/embed/embed.js (hunk 4 above)
Comment (sebastianbenz) [inline templates/embed/embed.js]: Of course! Good idea! done.

Comment (pbakaus) [inline templates/embed/embed.js]: Great!

File: lib/Preview.js [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+const slug = require('slug');
+const pygmentize = require('pygmentize-bundled');
+const minify = require('html-minifier').minify;
+const path = require('path');
+const mkdirp = require('mkdirp').sync;
+const phantom = require('phantom');
+
+const DocumentParser = require('./DocumentParser');
+
+const TEMPLATES_DIR = path.join(__dirname, '../templates/embed');
```
Comment (sebastianbenz) [inline lib/Preview.js]: done

File: lib/Preview.js (hunk 1 above)
Comment (sebastianbenz) [inline lib/Preview.js]: Can you install chrome via node?

File: templates/embed/embed.js (hunk 4 above)
Comment (sebastianbenz) [inline templates/embed/embed.js]: This made things so much easier...

File: templates/embed/embed.html (hunk 2 above)
Comment (sebastianbenz) [inline templates/embed/embed.html]: sgtm

**Comment by sebastianbenz** [comment]
PTAL

**Comment by sebastianbenz** [comment]
Sample here: https://amp-live-list-form.glitch.me/embeds.html

File: lib/Preview.js (hunk 1 above)
Comment (pbakaus) [inline lib/Preview.js]: AFAIK no, so suggested plan:

1. Install chrome via apt on Linux and Travis: https://iamakulov.com/notes/all/install-google-chrome-on-travis/
2. Follow these instructions to use via node: https://chromium.googlesource.com/chromium/src/+/lkgr/headless/README.md

Comment (pbakaus) [inline lib/Preview.js]: Also, this will come in handy if you go that route: https://chromedevtools.github.io/debugger-protocol-viewer/tot/DOM/

Comment (pbakaus) [inline lib/Preview.js]: also, via @paulirish (thx!), this doc should be super helpful as a frame of context: https://medium.com/@dschnr/using-headless-chrome-as-an-automated-screenshot-tool-4b07dffba79a

**Comment by pbakaus** [comment]
Nice, super smooth! Ship it! Conversion to headless Chrome can come at a later PR, and is optional I think.

**Comment by sebastianbenz** [comment]
Thanks for the review! I'm not sure the benefits of using headless chrome justify the additional setup required.

**Pull request merged**